docs/architecture/adr-056-proving-amnesia-attack.md

```diff
@@ -0,0 +1,117 @@
+
+```
+type ProofOfLockChange struct {
+	Votes []*types.Vote
+}
+```
+
+This can be either evidence of +2/3 PREVOTES or PRECOMMITS (either warrants the honest node the right to vote) and is valid, among other checks, so long as the PRECOMMIT vote of the node in V2 came after all the votes in the `ProofOfLockChange` i.e. it received +2/3  votes and then voted itself (F is unable to prove this).
+
+In the event that an honest node receives `PotentialAmnesiaEvidence` it will first `Verify()` it and then will check if it is among the suspected nodes in the evidence. If so, it will retrieve the `ProofOfLockChange` and combine it with `PotentialAmensiaEvidence` to form `AmensiaEvidence`:
+
+```
+type AmnesiaEvidence struct {
+	Evidence *types.PotentialAmnesiaEvidence
+	Proofs	 []*types.ProofOfLockChange
+}
+```
+
+If the node is not required to submit any proof than it will simply broadcast the `PotentialAmnesiaEvidence` .
+
+In either case, in a network where each node has n peers, if n is high enough to easily overcome network problems such as dropout than it should be sufficient that peers only broadcast once per `Amnesia trial period` + `Gossip safety margin` as long as `AmnesiaEvidence` has not yet been committed on the chain.
+
+If a node does not have sufficient peers then it may proportionally send more messages at an appropriate interval.
+
+When a node has successfully validated `PotentialAmnesiaEvidence` it timestamps it and refuses to receive the same form of `PotentialAmnesiaEvidence`. If a node receives `AmnesiaEvidence` it checks it against any current `AmnesiaEvidence` it might have and if so merges the two by adding the proofs, if it doesn't have it yet it run's `Verify()` and stores it.
+
+There can only be one `AmnesiaEvidence` and one `PotentialAmneisaEvidence` stored for each attack (i.e. for each height).
+
+When, `time.Now() > PotentialAmnesiaEvidence.timestamp + AmnesiaTrialPeriod`, honest validators of the current validator set can begin proposing the block that contains the `AmnesiaEvidence`.
+
+Other validators will vote <nil> if:
+
+- The Amnesia Evidence is not valid
+- The Amensia Evidence is not within the validators trial period - the gossip safety margin i.e. too soon.
+- The Amensia Evidence is of the same height but is different to the Amnesia Evidence that they have. i.e. is missing proofs.
+- Is of an AmnesiaEvidence that has already been committed to the chain.
+
+*NOTE: Evidence is capped at 484 bytes. `AmensiaEvidence`, unlike other evidence is batch processed and so we need to be careful as to not exceed this value or
+have a method of breaking the evidence first before sending it. An alternative is to find a way to have dynamic evidence size such that evidence is attached before the
+transactions are attached from the mempool.*
+
+*NOTE: Although this process only indicates committing `AmensiaEvidence` to the chain, it is also possible that `PotentialAmnesiaEvidence` is committed*
+
```

Comment:
I guess we should also highlight that handling evidence on chain is an optimistic approach. If +1/3 of the validators are faulty, they have enough votes to censor any evidence. Also if there is a real fork with some correct validators have decided on block A and other correct validators have decided on block B, we will need social consensus the establish a re-starting point.

Reply:
In a way `AmnesiaEvidence` is gossiped off-chain such that all nodes should eventually receive the evidence and perhaps without needing to commit it on-chain, halt and then use some sort of social consensus to restart. There was the notion of stopping evidence censorship here: #4593